lib/plugins/azurepush.js

```diff
@@ -0,0 +1,323 @@
+        // *****REMOVE*****: Beta logging
+        if (err) {
+          logTrelloIssue(env, '0x08-Failed to verify index on alert in mongodb');
+        }
+        // END: *****REMOVE*****
+        
+        // TODO: Handle error
+        console.log(err, doc);
+      });
+      ctx.store.collection(env.azurepush_collection).ensureIndex('settings.info', function indexAnnouncement(err, doc) {
+        // *****REMOVE*****: Beta logging
+        if (err) {
+          logTrelloIssue(env, '0x09-Failed to verify index on info in mongodb');
+        }
+        // END: *****REMOVE*****
+        
+        // TODO: Handle error
+        console.log(err, doc);
+      });
+      
+      // More info on request body: https://msdn.microsoft.com/en-us/library/mt621153.aspx
+      var installation = {
+        installationId: request.installationId
+        , pushChannel: request.deviceToken
+        , templates: {
+          nightscout: {
+            tags: tags
+          }
+        }
+      };
+
+      // 'platform': https://developer.xamarin.com/guides/xamarin-forms/platform-features/device/#Device.OS
+      switch (request.platform) {
+        case 'iOS':
+          installation.platform = 'apns';
+          installation.templates.nightscout.body = '{\"aps\":{\"alert\":\"$(message)\",\"sound\":\"$(sound)\",\"category\":\"$(category)\"},\"eventName\":\"$(eventName)\",\"group\":\"$(group)\",\"key\":\"$(key)\",\"level\":\"$(level)\",\"title\":\"$(title)\"}';
+          break;
+        case 'Android':
+          installation.platform = 'gcm';
+          installation.templates.nightscout.body = '{\"data\":{\"message\":\"$(message)\",\"eventName\":\"$(eventName)\",\"group\":\"$(group)\",\"key\":\"$(key)\",\"level\":\"$(level)\",\"sound\":\"$(sound)\",\"title\":\"$(title)\"}}';
+          break;
+        case 'WinPhone': //Not yet supported
+        case 'Windows': //Not yet supported
+        default: //No platform specified, or invalid platform
+          // *****REMOVE*****: Beta logging
+          logTrelloIssue(env, '0x10-Register received without supported platform');
+          // END: *****REMOVE*****
+
+          return callback({message: 'Platform not supported'});
+      }
+
+      notificationHubService.createOrUpdateInstallation(installation, function (error, response) {
+        // *****REMOVE*****: Beta logging
+        if (error) {
+          logTrelloIssue(env, '0x11-Failed to create or update installation');
+        }
+        // END: *****REMOVE*****
+
+        callback(error, {installationId: installation.installationId});
+      });
+    }
+    else {
+      // *****REMOVE*****: Beta logging
+      logTrelloIssue(env, '0x12-Register received without platform');
+      // END: *****REMOVE*****
+
+      // No platform, required parameter
+      return callback({message: 'Platform required'});
+    }
+  };
+
+  azurepush.ack = function ack(request, callback) {
+    if (request && request.level) {
```

Comment:
I thought about this quite a bit... and was uneasy about the openness of an API like this too.  You'll also note that there was a separate existing API that required the caller be authenticated with an API_SECRET.

The problem is we don't currently have a way to uniquely identify notifications beyond the group and level.  The key is currently being sent in both directions, but not being used since it's not unique to every notification.  It could be used and checked against the notify table, but it's not very secure (since some notification keys are as simple as ar2_1/ar2_2).
... As a side note, is it true that snoozing is currently open to any unauthenticated client on the website?

Additionally, there is currently no requirement to have an API_SECRET input on the client to use notifications (nor do I think there should be).

One thing we could do is use a unique _id generated to store notifications in mongo (which we currently don't do, but want to do at some point) that would allow us to easily persist a uuid and check on acknowledgement.  That wouldn't be too difficult.